sim/core/spell_mod.go

```diff
@@ -0,0 +1,272 @@
+package core
+
+import (
+	"strconv"
+	"time"
+)
+
+/*
+SpellMod implementation.
+*/
+
+type SpellModConfig struct {
+	ClassMask  int64
+	Kind       SpellModType
+	School     SpellSchool
+	IntValue   int64
+	TimeValue  time.Duration
+	FloatValue float64
+}
+
+type SpellMod struct {
+	ClassMask      int64
+	Kind           SpellModType
+	School         SpellSchool
+	floatValue     float64
+	intValue       int64
+	timeValue      time.Duration
+	Apply          SpellModApply
+	Remove         SpellModRemove
+	IsActive       bool
+	AffectedSpells []*Spell
+}
+
+type SpellModApply func(mod *SpellMod, spell *Spell)
+type SpellModRemove func(mod *SpellMod, spell *Spell)
+type SpellModFunctions struct {
+	Apply  SpellModApply
+	Remove SpellModRemove
+}
+
+func buildMod(unit *Unit, config SpellModConfig) *SpellMod {
+	functions := spellModMap[config.Kind]
+	if functions == nil {
+		panic("SpellMod " + strconv.Itoa(int(config.Kind)) + " not implmented")
+	}
+
+	mod := &SpellMod{
+		ClassMask:  config.ClassMask,
+		Kind:       config.Kind,
+		School:     config.School,
+		floatValue: config.FloatValue,
+		intValue:   config.IntValue,
+		timeValue:  config.TimeValue,
+		Apply:      functions.Apply,
+		Remove:     functions.Remove,
+		IsActive:   false,
+	}
+
+	unit.OnSpellRegistered(func(spell *Spell) {
+		if shouldApply(spell, mod) {
+			mod.AffectedSpells = append(mod.AffectedSpells, spell)
+
+			if mod.IsActive {
+				mod.Apply(mod, spell)
+			}
+		}
+	})
+
+	return mod
+}
+
+func (unit *Unit) AddStaticMod(config SpellModConfig) {
+	mod := buildMod(unit, config)
+	mod.Activate()
+}
+
+// Never use dynamic mods for Auras that have ExpireNever and activate on reset
+// Those mods will be overwritten potentilly during sim reset
+func (unit *Unit) AddDynamicMod(config SpellModConfig) *SpellMod {
+	return buildMod(unit, config)
+}
```

Comment:
Why is this a problem for NeverExpire and activation on reset?
Usually auras that are set to be permanent (aka neverexpire+activate on reset) are properly deactivated at the end of the iteration and activated at the start of the next.

Reply:
Yes. But before auras a disabled, we do Spell.Rest - and in spell.Reset certain parameters the aura modified are overwritten.

So it's basically like this right now.

Static Mods
Spell Snapshot (Initial Values)
Dynamic Mods
Spell Reset (Overwrite values like DamageMultiplier, DamageMultiplerAdd)
Aura Reset (Where we would remove mods we applied but those mods effects was removed in Spell Reset potentially)